test/test_multidevice_pipeline.cpp

```diff
@@ -596,59 +596,88 @@ TEST_F(PipelineTest, matmuls_megatron_mlp) {
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+    tv->setDeviceMesh(mesh);
+  }
 
-  // Matmul 1: tvx x tva -> y (N,M)
-  TensorView* tvx_b = broadcast(tvx, {true, false, false}); // (D_model, S, H*D)
-  TensorView* tvw_b = broadcast(tvw, {false, true, false}); // (D_model, S, H*D)
-  TensorView* tvxw = mul(tvx_b, tvw_b); // (D_model, S, H*D)
-  TensorView* tvy = sum(tvxw, {2}); // (D_model, S) y^T
-  TensorView* tvy_ = add(tvy, bias);
-  fusion->addOutput(tvy_);
-
-  // input and output tensors are not sharded 
-  // matmul weights are sharded along the vocabulary dimension (D_model)
-  // tvw->split(0, num_devices, false);
-  tvw->axis(0)->parallelize(ParallelType::DIDx); 
-  // tvx_b->split(0, num_devices, false);
-  tvx_b->axis(0)->parallelize(ParallelType::DIDx); 
-  // tvxw_b->split(0, num_devices, false);
-  tvw_b->axis(0)->parallelize(ParallelType::DIDx); 
-  // tvxw->split(0, num_devices, false);
-  tvxw->axis(0)->parallelize(ParallelType::DIDx); 
-
-  std::vector<TensorView*> tvs = {tvx, tvw, tvx_b, tvw_b, tvxw, tvy, bias, tvy_};
+  // All other tensorviews that are not sharded
+  std::vector<TensorView*> tvs = {y0_, y1, y2, y, bias};
   for (auto tv : tvs) {
     tv->setDeviceMesh(mesh);
   }
   
-  auto x = at::randn({S, H*D}, tensor_options);
-  auto wT = at::randn({D_model, H*D}, tensor_options);
-  auto zero = at::zeros({D_model, S}, tensor_options);
-  inputs = {x, shardInputTensor(wT, mesh, communicator->deviceId()), zero};
-  auto y = at::matmul(x, wT.transpose(0,1));
+  auto aten_q = at::randn({H, S, Dk}, tensor_options);
+  auto aten_k = at::randn({H, S, Dk}, tensor_options);
+  auto aten_v = at::randn({H, S, Dv}, tensor_options);
+  auto aten_w = at::randn({H, Dv, D_model}, tensor_options);
+  auto zero = at::zeros({S, D_model}, tensor_options);
+  auto deviceId = communicator->deviceId();
+  inputs = {shardInputTensor(aten_q, mesh, deviceId),
+            shardInputTensor(aten_k, mesh, deviceId),
+            shardInputTensor(aten_v, mesh, deviceId),
+            shardInputTensor(aten_w, mesh, deviceId),
+            zero};
+  auto aten_qk = at::matmul(aten_q, aten_k.permute({0, 2, 1})); // H, S, Dk x H, Dk, S -> H, S, S
+  auto aten_qkv = at::matmul(aten_qk, aten_v); // H, S, S x H, S, Dv -> H, S, Dv
+  auto aten_qkv_concat = aten_qkv.permute({1, 0, 2}).flatten(1, 2); // S, H*Dv 
+  auto aten_w_ = aten_w.flatten(0, 1); // H*Dv, Dmodel 
+  auto out = at::matmul(aten_qkv_concat, aten_w_); // S, H*Dv x H*Dv, Dmodel -> S, Dmodel
+  auto expected_outputs = {shardInputTensor(aten_qk, mesh, deviceId),
+                           shardInputTensor(aten_qkv, mesh, deviceId),
+                           out};
   
   MultiDeviceExecutor runtime(std::move(fusion), *communicator);
   auto outputs = runtime.runWithInput(inputs);
-    testValidate(
-        runtime.fusion(), outputs, inputs, {y.transpose(0, 1)}, __LINE__, __FILE__);
+  testValidate(
+        runtime.fusion(), outputs, inputs, expected_outputs, __LINE__, __FILE__);
 }
```

Comment:
I'm not sure we can validate this way. Doesn't it throw an error ? Indeed `at::randn` will give a different tensor at each rank, so it cannot work unless we gather the inputs at one rank. Alternative solution would be to give deterministic inputs

Why not using directly the `PipelineTest::validate` method?

Reply:
The tensors are the same on each process because they have the same random seed, but I agree it makes the test confusing. 
I tried using `executeAndValidate`, but got an exception from  getMatmulHeuristics. I think it's related to the executor cache problems, so I'll look into that.

C++ exception with description "mma_op.has_value() INTERNAL ASSERT FAILED at "Fuser/csrc/scheduler/matmul_utils.cpp":379, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. Failed to determine a MMA op for given problem.
Exception raised from getMatmulHeuristics at /Fuser/csrc/scheduler/matmul_utils.cpp:379 (most recent call first):

Reply:
> The tensors are the same on each process because they have the same random seed, but I agree it makes the test confusing.

Yes right!

Actually now that Im thinking about it, maybe we could get rid of the whole infrastructure of `executeAndValidate` and instead do it the way you're doing. It would require from the test case that the input are sharded from a tensor available uniformly by all the ranks, but thats a very much reasonable requirement. It would allow a much simpler test implementation which could also run way faster.

I just feel a bit embarassed to erase all the this infrastructure that was quite long to implement, but that's life! :sweat_smile: I don't want hold onto smth if it's not good 

Wdyt?

> I tried using executeAndValidate, but got an exception from getMatmulHeuristics. I think it's related to the executor cache problems, so I'll look into that.

Ok I see!
I don't know what this issue is exactly, but that looks indeed like the matmul scheduling that is complaining. It would be useful some point to figure out if this is cause through `FusionExecutor` or `FusionExecutorCache` (and I suspect it is `FusionExecutor`) and then ask Xiang what is the proper way to schedule the matmul to avoid this issue

Reply:
I think we can update `executeAndValidate` to use the unsharded input and keep the functionality that recompiles the fusion as a single device fusion?